File: src/infinicore/ops/sparse_scatter/sparse_scatter.cc
#include "infinicore/ops/sparse_scatter.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(SparseScatter);

SparseScatter::SparseScatter(Tensor output, const SpVec &input) {
    INFINICORE_ASSERT(input);
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(output, input->values(), input->indices());
    INFINICORE_GRAPH_OP_DISPATCH(output->device().getType(), output, input);
}

void SparseScatter::execute(Tensor output, const SpVec &input) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(SparseScatter, output, input);
}

Tensor sparse_scatter(const SpVec &input) {
    INFINICORE_ASSERT(input);
    auto output = Tensor::zeros({input->size()}, input->dtype(), input->device());
    sparse_scatter_(output, input);
    return output;
}

void sparse_scatter_(Tensor output, const SpVec &input) {
    SparseScatter::execute(output, input);
}

} // namespace infinicore::op

File: python/infinicore/ops/sparse_scatter.py
from infinicore.lib import _infinicore
from infinicore.spvec import SpVec
from infinicore.tensor import Tensor


def sparse_scatter(input: SpVec, *, out=None):
    if not isinstance(input, SpVec):
        raise TypeError("sparse_scatter expects a COO SpVec input")

    if out is None:
        return Tensor(_infinicore.sparse_scatter(input._underlying))

    _infinicore.sparse_scatter_(out._underlying, input._underlying)
    return out
